File: src/Board.cpp
#include "Board.hpp"

#include <cstdint>

#include "Player.hpp"
#include "Position.hpp"

namespace connect_four
{
    Board::Board(uint_least8_t const row_count, uint_least8_t const col_count):
        _row_count(row_count),
        _col_count(col_count),
        _cells(row_count * col_count)
    { }

    bool Board::is_inside(Position const position)
    {
        return
            position.row >= 0 &&
            position.col >= 0 &&
            position.row < _row_count &&
            position.col < _col_count;
    }

    bool Board::is_empty(Position const position)
    {
        return is_inside(position) && !_at(position).is_filled;
    }

    bool Board::is_filled(Position const position, Player const player)
    {
        return
            is_inside(position) &&
            _at(position).is_filled &&
            _at(position).player == player;
    }

    void Board::set(Position const position, Player const player)
    {
        _at(position) = Cell{player, true};
    }

    Cell & Board::_at(Position const & position)
    {
        auto index = position.row * _col_count + position.col;

        return _cells[static_cast<uint8_t>(index)];
    }
}

File: csharp/Board.cs
namespace ConnectFour;

public class Board
{
    private readonly int _rowCount;
    private readonly int _colCount;
    private Player?[] _cells;

    public Board(int rowCount, int colCount)
    {
        _rowCount = rowCount;
        _colCount = colCount;
        _cells = new Player?[_rowCount * _colCount];
    }

    public bool IsInside(Position position)
    {
        return
            position.Row >= 0 &&
            position.Col >= 0 &&
            position.Row < _rowCount &&
            position.Col < _colCount;
    }

    public bool IsEmpty(Position position)
    {
        return IsInside(position) && !_cells[Index(position)].HasValue;
    }

    public bool IsFilled(Position position, Player player)
    {
        return IsInside(position) && _cells[Index(position)] == player;
    }

    public void Set(Position position, Player player)
    {
        _cells[Index(position)] = player;
    }

    private int Index(Position position)
    {
        return position.Row * _colCount + position.Col;
    }

    // public override string ToString()
    // {
    //     var s = "";

    //     for (var row = 0; row < _rowCount; row += 1)
    //     {
    //         for (var col = 0; col < _colCount; col += 1)
    //         {
    //             var idx = row * _colCount + col;
    //             switch (_cells[row * _colCount + col])
    //             {
    //                 case Player.PLAYER_1:
    //                 s += "X  ";
    //                 break;
    //                 case Player.PLAYER_2:
    //                 s += "O  ";
    //                 break;
    //                 default:
    //                 s += $"{idx}  ";
    //                 break;
    //             }
    //         }

    //         s += "\n";
    //     }

    //     return s;
    // }
}
